### PR TITLE
Add DataDriftTrigger: supports one Evidently metric

### PR DESCRIPTION
**Pull request by jenny011**
This is a clean version of PR#367.
1. Add DataDriftTrigger class to supervisor. Supports one configurable Evidently metric. Launches drift detection every N new data points. Data used in detection are data trained in the previous trigger and all the untriggered new data.
2. Update Trigger interface. `Trigger.inform()` returns a Generator instead of List.
3. Add a generic ModelDownloader in supervisor.
4. Add example pipelines using DataDriftTrigger.
5. Add Evidently to pylint known third party.
6. Change ModelDownloader to embedding encoder utils. The downloader sets up and returns the model. The DataDriftTrigger owns the model. 

Future
1. Support multiple configurable Evidently metric. #416
2. Support Alibi-Detect. #414 
3. Support custom embedding encoder. #417
4. Support different windowing for detection data, e.g. compare with all previously trained data. #418
5. Common DataLoaderInfo #415

### CODE REVIEW

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:206]: Why only log this if there are no triggers?

File: modyn/supervisor/internal/triggers/trigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/trigger.py:11]: I think we can make this a non-abstract method with just a `pass` implementation, since only the drift trigger uses this. Then we don't need the empty overrides in the other triggers.

File: modyn/supervisor/internal/triggers/trigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/trigger.py:30]: Same comment as above - can probably be non abstract methods to avoid empty overrides

File: modyn/supervisor/internal/triggers/model_downloader/model_downloader.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/model_downloader/model_downloader.py:23]: 366 might be the wrong issue number. I guess it would be cleaner to add this downloader as a common utility beforehand, but we can also do it afterwards. But we should definitly do it :)

File: modyn/supervisor/internal/triggers/model_downloader/model_downloader.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/model_downloader/model_downloader.py:87]: Why this assert? I think not every model has to be a coreset supporting moudle, or?

File: modyn/supervisor/internal/triggers/model_downloader/model_downloader.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/model_downloader/model_downloader.py:24]: This is a ModelDownloader + a class representing the downloaded model. To make the object oriented abstraction a bit cleaner, could we instead just `return` a model at the end? From my understanding, the model downloader should download + instantiate, but not _own_ the model. That would be cleaner. Thank you!

File: modyn/supervisor/internal/triggers/trigger_datasets/dataloader_info.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/trigger_datasets/dataloader_info.py:4]: I think that might be the wrong issue number. Might this be 275 instead?

File: modyn/supervisor/internal/triggers/trigger_datasets/online_trigger_dataset.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/trigger_datasets/online_trigger_dataset.py:12]: Could you add a docstring here explaining the idea of this dataset (sampling with some probability) and why we need it)?

File: modyn/supervisor/internal/triggers/trigger_datasets/fixed_keys_dataset.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/trigger_datasets/fixed_keys_dataset.py:25]: Can you add a docstring here what this does and why we need it? For the sampling dataset, I remember, but here some docs would be great. Thank you!

File: modyn/supervisor/internal/triggers/utils.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/utils.py:19]: I think this needs an introductory sentence on what this function is doing (like instantiating an evidently metric)

Comment (MaxiBoether) [reply modyn/supervisor/internal/triggers/utils.py:19]: Also, if you want to do a separate PR for your TODO 5 (which i think applies here, right), we might want to open an issue and link it here

File: modyn/supervisor/internal/triggers/utils.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/utils.py:134]: I think it would be cleaner to pass the tensor here and have a convert tensor to evidently embedding function instead

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/datadrifttrigger.py:200]: Might need a new issue here

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/datadrifttrigger.py:215]: I think this needs a longer docstring, sorry. Could you provide some longer documentation here, especially on the sliding window approach and what is happening? Right now, the docstring mostly concerns the return type

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/datadrifttrigger.py:231]: Should we move this at the end of the loop? Semenatically, we have not visited any new points here, have we? (Generally, I think some description of the algorithm as mentioned in the docstring comment would be super helpful)

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/datadrifttrigger.py:173]: Is this for logging/debugging purposes?

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/datadrifttrigger.py:177]: Can you also add a docstring here with some more context (e.g., we always compare against the previous trigger in the current implementation, right?)?

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/datadrifttrigger.py:155]: Can you add a comment on what the column mapping name is?

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thanks Jingyi! I did not have big comments, mostly smaller code-style changes and documentation issues. I think it might still be sensible to merge the downloader model thing first instead of extracting it later (probably overall less work) but we can do it however you prefer. As mentioned on Element, let's discuss how you proceed with the other TODOs. We definitely need the unit tests in this PR before merging :)

File: modyn/supervisor/internal/triggers/model_downloader/model_downloader.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/model_downloader/model_downloader.py:87]: Now, only the data drift trigger uses the model downloader, so the models should be able to record embeddings. Will deal with it in the general model downloader PR.

File: modyn/supervisor/internal/triggers/trigger_datasets/dataloader_info.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/trigger_datasets/dataloader_info.py:4]: 275 is about a common Dataset class, not DataLoaderInfo. I can create an issue about a common DataLoaderInfo.

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/datadrifttrigger.py:173]: Was used for experiments. It logs all drift detection metrics and results. Can remove it or make it debug-only.

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (jenny011) [reply modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:206]: This is the same as before supervisor refactorization. If there are triggers, the log is added inside _handle_new_data_batch().

File: modyn/supervisor/internal/triggers/trigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/trigger.py:11]: done

File: modyn/supervisor/internal/triggers/trigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/trigger.py:30]: done

File: modyn/supervisor/internal/triggers/model_downloader/model_downloader.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/model_downloader/model_downloader.py:23]: ModelDownloader Issue #433.

File: modyn/supervisor/internal/triggers/model_downloader/model_downloader.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/model_downloader/model_downloader.py:24]: The current model class does not store additional info as attributes. E.g. amp, device, device type. I have this model downloader class as a unified interface to access such additional info required to run the forward pass. A better way would probably be to implement an abstract model class that contains those additional info. I think it would better be dealt with in the ModelDownloader Issue #433. What do you think?

File: modyn/supervisor/internal/triggers/trigger_datasets/online_trigger_dataset.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/trigger_datasets/online_trigger_dataset.py:12]: done.

File: modyn/supervisor/internal/triggers/trigger_datasets/fixed_keys_dataset.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/trigger_datasets/fixed_keys_dataset.py:25]: done.

File: modyn/supervisor/internal/triggers/utils.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/utils.py:19]: Comment updated. Links to issues see the PR description.

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/datadrifttrigger.py:199]: Issue #417

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/datadrifttrigger.py:214]: done.

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/datadrifttrigger.py:176]: done

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/datadrifttrigger.py:154]: done

File: modyn/supervisor/internal/triggers/datadrifttrigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/datadrifttrigger.py:230]: It must be handled at the top because it is including all the remaining untriggered samples from the previous new data batch. The remaining untriggered samples have been processed in the previous batch, so there's no need to process them again. Explained in the docstring.

File: modyn/supervisor/internal/triggers/utils.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/utils.py:134]: Removed get_embeddings_evidently_format(). Added convert_tensor_to_df(tensor, column_name_prefix).

**Comment by MaxiBoether** [comment]
Thanks for the work. Let me know when I should review again.

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (XianzheMa) [inline modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:323]: just nitpicking: I think `num_triggers` is not needed as we can replace every occurrence of it with `len(triggering_idx_list)` right?

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (XianzheMa) [inline modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:308]: why do we remove this?
I'd like to keep it because this position is immediately after a trigger training / evaluation. We can persist the training / evaluation data so long as we have it.

File: modyn/supervisor/internal/triggers/amounttrigger.py
Comment (XianzheMa) [inline modyn/supervisor/internal/triggers/amounttrigger.py:20]: Why do we make it a `Generator[int, None, None]` instead of `Generator[int]`?

Comment (XianzheMa) [inline modyn/supervisor/internal/triggers/amounttrigger.py:20]: Why do we make it a `Generator[int, None, None]` instead of `Generator[int]`?

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (XianzheMa) [inline modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:323]: `num_triggers` is not needed as we can replace every occurrence of it with `len(triggering_idx_list)`.

It's not about nitpicking: better to remove it to not confuse others with `self.num_triggers`, which is a different number

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (XianzheMa) [inline modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:308]: why do we remove this?
I'd like to keep it because this position is immediately after a trigger training / evaluation. We can persist the training / evaluation data so long as we have it.

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (XianzheMa) [reply modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:183]: Thanks for keeping me in the loop. It looks good to me except for small changes!

**Review by XianzheMa — APPROVED** [review]
LGTM and can you address the small changes I requested before merging! Thanks!

File: modyn/supervisor/internal/triggers/trigger.py
Comment (jenny011) [reply modyn/supervisor/internal/triggers/trigger.py:11]: oops, missed it. Now it's fixed.

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (jenny011) [reply modyn/supervisor/internal/pipeline_executor/pipeline_executor.py:323]: done.

File: modyn/supervisor/internal/pipeline_executor/pipeline_executor.py
Comment (jenny011) [reply modyn/supervisor/internal/pipeline_executor/pipeline_executor.py]: It's probably lost in merging. Added back.

**Pull request merged**
